### ISSUE TITLE
508-defect-1 ❗ Launchblocker  [SCREENREADER, KEYBOARD]:  Map: All elements that can be operated by a mouse MUST be operable by keyboard

### ISSUE THREAD

**Issue by jenstrickland** [department-of-veterans-affairs/va.gov-team#6911]
# [508-defect-1 ❗️ Launchblocker](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-1)

```diff
! Leaflet-Mapbox issue
```

**Feedback framework**
- **❗️ Must** for if the feedback must be applied 
- **⚠️Should** if the feedback is best practice 
- **✔️ Consider** for suggestions/enhancements 

## Description

Per the VA 508 Office:
> (From 07/03/2019) Instructions are not provided that indicate interactive map keyboard shortcuts. For example, the user is not instructed that “-“ and “+” can zoom the map. (Not Resolved 3/11/2020) Unable to get to the map, and when the mouse is used there is no indication of the shortcuts (listed as a Critical issue)

- Related to the above: https://github.com/department-of-veterans-affairs/va.gov-team/issues/6104 - [AXE-CORE]: Zoom In / Zoom Out controls MUST have their visible text as part of their accessible name

- The only parts of the map that are keyboard accessible today are the **+** and **–**, and the links in the attribution. If focus is on either the **+** and **–** and a user selects enter or space to activate, it will create a change in the map, but there is no communicated change.

- When a mouse is used to access parts of the map itself there is nothing indicating what will happen, and no keyboard shortcuts are indicated that provide equitable interaction access.

There seems to be no great way, at this moment in time, for visually impaired and blind users to interact with maps. People who use screen readers, however, can interact with tables of data. Presenting the search results content semantically and offer the map as something optional may be a more accessible option. The map **MUST** provide keyboard shortcuts for interactions, be accessible by keyboard, and relay context change updates.

## Point of Contact

**VFS Point of Contact:** Jennifer

## Acceptance Criteria

- As a keyboard user, I want to know what the keyboard shortcuts are to interact with the map. 

## Environment

* Operating System: all
* Browser: all
* Server destination: production

## Steps to Recreate

1. Enter `https://www.va.gov/find-locations` in browser
2. Without using the mouse or trackpad, navigate the page, interact with the map
3. Verify that it is hard to understand how to zoom in/out and impossible to interact with the map itself
4. Additionally, recreate the above three steps with your screen reader, as screen reader users are often keyboard users as well.

## Possible Fixes (optional)

Since https://www.va.gov/directory/guide/home.asp is 508 compliant, I looked at that as a guide. That map is not really interactive, as it is solely an image map with each state serving as a link to a page of that state's locations.

The Carnegie Museums link below indicates many visually impaired users use Google Maps' mobile app quite satisfactorily, but the use case is different. Perhaps the mobile app interaction patterns may lead to solution ideas.

My initial inclination, which I sent to the VA 508 Office for guidance:
> This all leads me to think we may need to follow Google Maps' and Mapquest desktop behavior and skip map interactions for keyboard and screen readers. My first inclination is to suggest teams present a list of search results and offer to show a map only upon user interaction. Yet, I'm not sure what to recommend for the map page. Are you aware of any other maps on VA.gov that pass a review by your office that I might recommend the team model after?

I'll update with their response.

Any solutions should meet the following:

* Provide instructions that indicate interactive map keyboard shortcuts. 
* Ensure keyboard users are able to get to the map.
* Ensure there is an indication of the keyboard shortcuts provided when the mouse is used.
* Ensure context changes are communicated, accessible to screen readers.

## WCAG or Vendor Guidance (optional)

* [Understanding WCAG 2.0 - Keyboard](https://www.w3.org/TR/UNDERSTANDING-WCAG20/keyboard-operation-keyboard-operable.html)

* [Understanding WCAG 2.0 - On Focus, change of context](https://www.w3.org/TR/UNDERSTANDING-WCAG20/consistent-behavior-receive-focus.html)

* [Creating accessible maps, Carnegie Museums reference](http://web-accessibility.carnegiemuseums.org/content/maps/)

* [Techniques for WCAG 2.0 - G202: Ensuring keyboard control for all functionality](https://www.w3.org/TR/2016/NOTE-WCAG20-TECHS-20161007/G202)

**Comment by mmiddaugh** [comment]
Should be resolved by https://github.com/department-of-veterans-affairs/va.gov-team/issues/17963

**Comment by mmiddaugh** [comment]
Hi @joshkimux we believe these issues may be resolved. Can you please validate, when you are able?

**Comment by joshkimux** [comment]
@mmiddaugh Thank you for tagging me! This is definitely an improvement! Upon using either zoom buttons and `space` , the "search this area of the map" button is focused and announced to users.

That being said, there's still some bugs existing that will need to get ironed out. It may be easier to talk these through over a call. I'll do my best to document them in this comment just to set the context for such a discussion.

## The image of the map is focusable, but provides no indication of focus
Before I give any feedback here, I'm curious to learn if this was intended behavior-- and if so, what does it mean to achieve/what value does it provide? If it was not intended, I'd recommend removing it from the tab order. Otherwise, keyboard users without a screen reader will have an invisible tab stop.

<img width="1111" alt="Screen Shot 2021-01-06 at 9 42 24 AM" src="https://user-images.githubusercontent.com/14154792/103781793-c9c2a500-5004-11eb-80a1-c650451171d3.png">

## Error messaging is not specific
There are a couple of launch blocker issues here that I think can be fixed relatively easily. 
1. The error message itself is not specific. Screen reader users won't know _what_ exactly is the problem if the problem is called "this" instead of say "Fill in a city, state, or postal code."
2. Errors are tooltips instead of being persistent within the label of the field itself. Was there a design/engineering constraint that resulted in this vs. the traditional form of error validation?

https://user-images.githubusercontent.com/14154792/103782570-9f251c00-5005-11eb-955d-8a1a45420f50.mov
<img width="661" alt="Screen Shot 2021-01-06 at 9 41 24 AM" src="https://user-images.githubusercontent.com/14154792/103782591-a6e4c080-5005-11eb-957c-eaad37f44616.png">

## Changes of context when zooming are not announced to screen reader users
When a screen reader user presses the zoom in/out buttons, no feedback is provided to them that the map has indeed zoomed in or out. No feedback is provided on if more or less items have appeared onto the map either (which may be technically infeasible to do). This will cause some issues when combined with the following issue:

## Zoom in/out to Search this area button focus management is inconsistent
Couple of problems here that will need a deeper discussion on:

- On Chrome, if a user fires a zoom in/out event for the first time using `space` or `enter`, the Search this area button will appear and receive focus. This behavior is acceptable from an accessibility standpoint, but from a UX perspective it can be argued that this is taking control away from the user. e.g. What if a user wants to zoom out several times before conducting a search? A polite notification that the option to search this area has appeared OR making the search this area button always persistent may address this.
- On Safari/VoiceOver, the `space` or `enter` keys were inconsistent in moving focus to the Search this area button. Although focus sometimes _would not_ be moved, the screen reader still announced the Search this area button as though focus _had_ been moved. When compounded with the fact that changes of context when zooming are not announced, a user may continue hitting the zoom in/out buttons _thinking_ they are searching the area when in reality they're still on the zoom in/out buttons.
- On all browsers, the search this area button is not reachable in the tab order. It is only accessible when focus is moved to it on the _first time_ the user fires a zoom in or out.

I'd recommend we meet to talk about this last one because it's a little complicated 🙇

**Comment by mmiddaugh** [comment]
Thanks for your thorough feedback, @joshkimux. We have a time set up for Jan 11 to discuss many of these items with you. 

**Relevant issues** 
- The image of the map is focusable, but provides no indication of focus
- Zoom in/out to Search this area button focus management is inconsistent [#17665](https://github.com/department-of-veterans-affairs/va.gov-team/issues/17665)
- Error messaging is not specific [#18208](https://github.com/department-of-veterans-affairs/va.gov-team/issues/18208)
- Changes of context when zooming are not announced to screen reader users 
  -[#17448](https://github.com/department-of-veterans-affairs/va.gov-team/issues/17448)
  -[#6104](https://github.com/department-of-veterans-affairs/va.gov-team/issues/6104)
  -[#17656](https://github.com/department-of-veterans-affairs/va.gov-team/issues/17656)

**Comment by mmiddaugh** [comment]
Following the conversation with Josh ([recording](https://drive.google.com/file/d/1drsmPly33CQHGzVsxrvY_0y9OmMiHrZR/view?usp=sharing)), I will close this issue knowing the relevant items are covered in the following
- The image of the map is focusable, but provides no indication of focus/inconsistent focus [#17665](https://github.com/department-of-veterans-affairs/va.gov-team/issues/17665)
- Error messaging is not specific [#18351](https://github.com/department-of-veterans-affairs/va.gov-team/issues/18351)
- Changes of context when zooming are not announced to screen reader users [#17963](https://github.com/department-of-veterans-affairs/va.gov-team/issues/17963)
